File: PageNamePlayground.playground/Sources/PageName.swift
import Foundation

public struct SDKSettings {
    public let appRoot: String
    public init(appRoot: String){
        self.appRoot = appRoot
        SDKSettings.Shared = self
    }
    
    public static private(set) var  Shared: SDKSettings?
}

public struct PageName {
    private var _parts: [String] = []
    
    var parts:[String] {
        _parts
    }
    private let _prefix: String
    
    public init() {
        // I kinda want to do this for each new page name BUT i don't want to do it when adding page names together
        _prefix = SDKSettings.Shared?.appRoot ?? "Not Set"
        _parts.append(_prefix)
    }
    
    public init(_ pageNames: String...) {
        _prefix = SDKSettings.Shared?.appRoot ?? "Not Set"
        _parts.append(_prefix)
        for pageName in pageNames {
            _parts.append(pageName)
        }
    }
    
    public init(_ pageName: PageName) {
        _prefix = SDKSettings.Shared?.appRoot ?? "Not Set"
        _parts.append(_prefix)
        append(pageName)
    }
    
    public mutating func append(_ part: String) {
        _parts.append(part)
    }
    
    public mutating func append(contentsOf: [String]) {
        _parts.append(contentsOf: contentsOf)
    }
    
    public mutating func append(_ contentsOf: PageName) {
        append(contentsOf: contentsOf.parts.filter { $0 != _prefix})
    }
    
    public mutating func pop() -> String? {
        _parts.popLast()
    }
}

extension PageName: CustomStringConvertible {
    public var description: String {
        self.parts.joined(separator: ":")
    }
}

extension PageName {
    
    public static var Prefix: PageName {
        PageName()
    }
    
    public static func +=(lhs: inout PageName, rhs: PageName) -> PageName {
        lhs.append(rhs)
        return lhs
    }
    
    public static func +=(lhs: inout PageName, rhs: String) -> PageName {
        lhs.append(rhs)
        return lhs
    }
    
    public static func +(lhs: PageName, rhs: PageName) -> PageName {
        var lhs = PageName(lhs)
        lhs.append(rhs)
        return lhs
    }
    
    public static func +(lhs: PageName, rhs: String) -> PageName {
        var lhs = PageName(lhs)
        lhs.append(rhs)
        return lhs
    }
}

File: kotlin-pagename/src/main/kotlin/PageName.kt
class SDKSettings (val appRootName: String) {
    init {
        Shared = this;
    }
    companion object {
        var Shared: SDKSettings? = null
            private set
    }
}
class PageName {
    private var _parts: MutableList<String> = mutableListOf()
    // Kinda hate this
    // I hate this shared pattern
    private val _prefix: String = SDKSettings.Shared?.appRootName ?: "Not Set"

    constructor(vararg  pageNameParts: String) {
        _parts.add(_prefix)
        for (part in pageNameParts) {
            add(part)
        }
    }

    fun add(part: String): Boolean {
        if(part == _prefix){ return false }
        return _parts.add(part)
    }

    fun pop(): String {
        return _parts.removeLast()
    }

    override fun toString(): String {
        return _parts.joinToString(":")
    }
}